service/history/api/consistency_checker.go

```diff
@@ -178,9 +179,91 @@ func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByClock(
 		release(err)
 		return nil, err
 	}
+
 	return NewWorkflowLease(wfContext, release, mutableState), nil
 }
 
+func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
+	ctx context.Context,
+	namespaceID string,
+	workflowID string,
+	lockPriority workflow.LockPriority,
+) (WorkflowLease, error) {
+	runID, err := c.GetCurrentRunID(
+		ctx,
+		namespaceID,
+		workflowID,
+		lockPriority,
+	)
+	if err != nil {
+		return nil, err
+	}
+	workflowLease, err := c.getWorkflowLease(
+		ctx, definition.NewWorkflowKey(namespaceID, workflowID, runID), lockPriority,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if workflowLease.GetMutableState().IsWorkflowExecutionRunning() {
+		return workflowLease, nil
+	}
+
+	currentRunID, err := c.GetCurrentRunID(ctx, namespaceID, workflowID, lockPriority)
+
+	if err != nil {
+		workflowLease.GetReleaseFn()(err)
+		return nil, err
+	}
+	if currentRunID == workflowLease.GetContext().GetWorkflowKey().RunID {
+		return workflowLease, nil
+	}
+
+	workflowLease.GetReleaseFn()(nil)
+	return nil, consts.ErrLocateCurrentWorkflowExecution
+}
+
+// The code below is an old implementation of GetWorkflowLease
+// It should be avoided. We should revisit the logic
+
+func (c *WorkflowConsistencyCheckerImpl) GetWorkflowLeaseWithConsistancyCheck(
+	ctx context.Context,
+	reqClock *clockspb.VectorClock,
+	consistencyPredicate MutableStateConsistencyPredicate,
+	workflowKey definition.WorkflowKey,
+	lockPriority workflow.LockPriority,
+) (WorkflowLease, error) {
+	if reqClock != nil {
+		currentClock := c.shardContext.CurrentVectorClock()
+		if vclock.Comparable(reqClock, currentClock) {
+			return c.getWorkflowLeaseValidatedByClock(
+				ctx,
+				reqClock,
+				currentClock,
+				workflowKey,
+				lockPriority,
+			)
+		}
+		// request vector clock cannot is not comparable with current shard vector clock
+		// use methods below
```

Comment:
Can we have only one copy of logic for `reqClock != nil` case?

Reply:
in the first iteration the idea was to keep old code intact.
Yes, if we fine - I will combine the code.